cuqi/problem/_problem.py

```diff
@@ -1,64 +1,123 @@
+from __future__ import annotations
 import numpy as np
 import time
+from typing import Tuple
 
 import cuqi
 from cuqi import config
-from cuqi.distribution import Cauchy_diff, GaussianCov, InverseGamma, Laplace_diff, Gaussian, GMRF, Lognormal, Posterior, LMRF, Laplace, Beta
+from cuqi.distribution import Distribution, GaussianCov, InverseGamma, Laplace_diff, Gaussian, GMRF, Lognormal, Posterior, LMRF, Beta, JointDistribution
+from cuqi.density import Density
 from cuqi.model import LinearModel, Model
+from cuqi.likelihood import Likelihood
 from cuqi.geometry import _DefaultGeometry
 from cuqi.utilities import ProblemInfo
-from cuqi.pde import SteadyStateLinearPDE
+from cuqi.samples import CUQIarray
 
 from copy import copy
 
-
 class BayesianProblem(object):
-    """Representation of a Bayesian inverse problem (posterior) defined by a likelihood and prior.
-    
-    .. math::
+    """ Representation of a Bayesian inverse problem defined by any number of densities (distributions and likelihoods), e.g.
 
-        \pi_\mathrm{posterior}(\mathbf{x} \mid \mathbf{b}) \propto \pi_\mathrm{likelihood}(\mathbf{b} \mid \mathbf{x}) \pi_\mathrm{prior}(\mathbf{x}),
+    .. math::
 
-    where :math:`\pi_\mathrm{Likelihood}(\mathbf{b} \mid \mathbf{x})` is a :class:`cuqi.likelihood.Likelihood` function and :math:`\pi_\mathrm{prior}(\mathbf{x})` is a :class:`cuqi.distribution.Distribution`.
+        \\begin{align*}
+        \mathrm{density}_1 &\sim \pi_1 \\newline
+        \mathrm{density}_2 &\sim \pi_2 \\newline
+                           &\\vdots
+        \end{align*}
 
-    The main goal of this class is to provide fully automatic methods for computing samples or point estimates of the posterior distribution.
+    The main goal of this class is to provide fully automatic methods for computing samples or point estimates of the Bayesian problem.
 
     Parameters
     ----------
-    likelihood : Likelihood
-        The likelihood function.
+    \*densities: Density
+        The densities that represent the Bayesian Problem.
+        Each density is passed as comma-separated arguments.
+        Can be Distribution, Likelihood etc.
+
+    \**data: ndarray, Optional
+        Any potential observed data. The data should be passed
+        as keyword arguments, where the keyword is the name of the
+        density that the data is associated with.
+        Data can alternatively be set using the :meth:`set_data` method,
+        after the problem has been initialized.
+
+    Example
```

Comment:
This is a nice but comprehensive example. I think a super short - and incomplete - example would be helpful just to get the syntax across, e.g. something like
```
BayesianProblem(x, y, z, y=y_data)
```
or equivalently 
```
BayesianProblem(x, y, z).set_data(y=y_data)
```
for distributions x, y, z and observed data y_data for y.

Reply:
Good suggestion. I added a basic syntax example.